AppController/djinn.rb

```diff
@@ -4731,6 +4616,66 @@ def starts_new_apps_or_appservers()
     }
   end
 
+  # This function compares the applications that should be running with
+  # the ones we have setup, and removes the stopped applications.
+  def check_stopped_app()
+    uac = UserAppClient.new(my_node.private_ip, @@secret)
+    Djinn.log_debug("Checking applications that have been stopped.")
+    begin
+      app_list = uac.get_all_apps()
+    rescue FailedNodeException
+      Djinn.log_warn("Failed to get app listing: retrying.")
+      retry
+    end
+    app_list += HelperFunctions.get_loaded_apps()
+    app_list.each{ |app|
+      next if @app_names.include?(app)
+      next if RESERVED_APPS.include?(app)
+      begin
+        next if uac.is_app_enabled?(app)
+      rescue FailedNodeException
+        Djinn.log_warn("Failed to talk to the UserAppServer about " +
+          "application #{app}")
+        next
+      end
+
+      Djinn.log_info("#{app} is no longer running: removing old states.")
+      Djinn.log_run("rm -rf #{HelperFunctions.get_app_path(app)}")
+      CronHelper.clear_app_crontab(app)
+      Djinn.log_debug("(stop_app) Maybe stopping taskqueue worker")
+      maybe_stop_taskqueue_worker(app)
+      Djinn.log_debug("(stop_app) Done maybe stopping taskqueue worker")
+
+      APPS_LOCK.synchronize {
+        if my_node.is_load_balancer?
+          stop_xmpp_for_app(app)
+          Nginx.remove_app(app)
+          HAProxy.remove_app(app)
+        end
+
+        if my_node.is_appengine?
+          Djinn.log_debug("(stop_app) Calling AppManager for app #{app}")
+          app_manager = AppManagerClient.new(my_node.private_ip)
+          begin
+            if app_manager.stop_app(app)
+              Djinn.log_info("(stop_app) AppManager shut down app #{app}")
+            else
+              Djinn.log_error("(stop_app) unable to stop app #{app}")
+            end
+          rescue FailedNodeException
+            Djinn.log_warn("(stop_app) #{app} may have not been stopped")
```

Comment:
With the current implementation of the AppManagerServer, I'm not sure we can be more specific. Once we have a more rich way to interact with it, we will be able to get rid of this.  Since this can be a timeout, the state of the app would be unknown. We do check if monit reports it running next time around, so at least we will catch it eventaully.